src/openassetio-python/cmodule/CMakeLists.txt

```diff
@@ -124,3 +124,102 @@ if (OPENASSETIO_ENABLE_TESTS)
     target_compile_definitions(openassetio-python-module PRIVATE OPENASSETIO_ENABLE_TESTS)
     target_link_libraries(openassetio-python-module PRIVATE openassetio-python-module-test)
 endif ()
+
+
+#-----------------------------------------------------------------------
+# Generate .pyi stubs
+#
+# pybind11-stubgen will create openassetio/_openassetio subdirectories
+# in the build output directory. We use this knowledge to assemble a
+# Python pseudo-package in the build output directory, such that stubgen
+# can import the module to do its thing then dump its output alongside.
+
+if (OPENASSETIO_ENABLE_PYTHON_STUBGEN)
+    # Check pybind11-stubgen is available.
+    execute_process(
+        COMMAND "${Python_EXECUTABLE}" -m pybind11_stubgen --help
+        OUTPUT_VARIABLE _pybind11_stubgen_output
+        ERROR_VARIABLE _pybind11_stubgen_output
+        RESULT_VARIABLE _pybind11_stubgen_exit_code
+    )
+
+    # Fatal error if pybind11-stubgen isn't available.
+    if (NOT _pybind11_stubgen_exit_code EQUAL "0")
+        message(
+            FATAL_ERROR
+            "OPENASSETIO_ENABLE_PYTHON_STUBGEN=${OPENASSETIO_ENABLE_PYTHON_STUBGEN} but"
+            " pybind11-stubgen not found: status=${_pybind11_stubgen_exit_code}:"
+            " ${_pybind11_stubgen_output}"
+        )
+    endif()
```

Comment:
[Consider] Should this error be fatal? Not being able to generate stubs feels like a warning sort of thing for me.

Reply:
It's tricky.  I wrestled with the best error/warning behaviour.  

In particular I had in mind that stubgen is enabled by default, so people cloning the repo and building from source may encounter this.  

So the first question is, should it be enabled by default?  I reasoned yes, because our default config is deliberately the same as the config used for a release artefact.

Given it's enabled by default because our default config is for a releasable build, should it be a fatal error if stubgen isn't available? I reasoned yes, because otherwise we (or someone else) could easily accidentally build and release without stubs, e.g. if the logs are hidden behind CI.

There is a hole in our tests for Python wheel builds, since the stubgen tests are skipped if stubs are not found and there is no env var saying they should exist.  For Python wheel builds the env var isn't set (since it's set inside of CMake and only applies to `ctest`).

I guess we could set that env var separately in CI during wheel builds, to force the stubgen tests to run.  That would mitigate somewhat the hole in our tests :thinking: ....?

Reply:
> because otherwise we (or someone else) could easily accidentally build and release without stubs, e.g. if the logs are hidden behind CI.

Damn fine point!

Reply:
Forced stubgen tests to run in 6fc8db96

Added a drop! commit to illustrate - resulting workflow run: https://github.com/OpenAssetIO/OpenAssetIO/actions/runs/9127570146/job/25098125757?pr=1306 - now removed drop! commit.

Reply:
Renamed the environment variable in 60e6779a since now that it can be provided "from outside" it made it look like it would set the CMake variable, when it doesn't, it's only used in tests.